packages/apollo-engine-reporting/src/schemaReporter.ts

```diff
@@ -0,0 +1,166 @@
+        sendNextWithExecutableSchema = withExecutableSchema;
+        setTimeout(inner, inSeconds * 1000);
+      })
+      .catch((error: any) => {
+        // In the case of an error we want to continue looping
+        // We can add hardcoded backoff in the future,
+        // or on repeated failures stop responding reporting.
+        logger.error(
+          `Error in reporting server info to Apollo Graph Manager for schema reporting: ${error}`,
+        );
+        sendNextWithExecutableSchema = false;
+        setTimeout(inner, fallbackReportingDelayInMs);
+      });
+  }
+
+  inner();
+}
+
+interface ReportServerInfoReturnVal {
+  inSeconds: number;
+  withExecutableSchema: boolean;
+}
+
+// This class is meant to be a thin shim around the gql mutations.
+export class SchemaReporter {
+  // These mirror the gql variables
+  private readonly serverInfo: EdgeServerInfo;
+  private readonly executableSchemaDocument: any;
+  private readonly url: string;
+
+  private isStopped: boolean;
+  private readonly headers: Headers;
+
+  constructor(
+    serverInfo: EdgeServerInfo,
+    schemaSdl: string,
+    apiKey: string,
+    schemaReportingEndpoint: string | undefined,
+  ) {
+    this.headers = new Headers();
+    this.headers.set('Content-Type', 'application/json');
+    this.headers.set('x-api-key', apiKey);
+    this.headers.set('apollographql-client-name', 'apollo-engine-reporting');
+    this.headers.set(
+      'apollographql-client-version',
+      require('../package.json').version,
+    );
+
+    this.url =
+      schemaReportingEndpoint ||
+      'https://engine-graphql.apollographql.com/api/graphql';
+
+    this.serverInfo = serverInfo;
+    this.executableSchemaDocument = schemaSdl;
+    this.isStopped = false;
+  }
+
+  public stopped(): Boolean {
+    return this.isStopped;
+  }
+
+  public stop() {
+    this.isStopped = true;
+  }
+
+  public async reportServerInfo(
+    withExecutableSchema: boolean,
+  ): Promise<ReportServerInfoReturnVal> {
+    const { data, errors } = await this.graphManagerQuery({
+      info: this.serverInfo,
+      executableSchema: withExecutableSchema
+        ? this.executableSchemaDocument
+        : null,
+    } as ReportServerInfoVariables);
+
+    if (errors) {
+      throw new Error((errors || []).map((x: any) => x.message).join('\n'));
+    }
+
+    if (!data || !data.me || !data.me.__typename) {
+      throw new Error(`
+Heartbeat response error. Received incomplete data from Apollo Graph Manager.
+If this continues please reach out at support@apollographql.com.
+Got response: "${JSON.stringify(data)}"
+      `);
+    }
+
+    if (data.me.__typename === 'UserMutation') {
+      this.isStopped = true;
+      throw new Error(`
+      This server was configured with an API key for a user.  Only a service's API key may be used for schema reporting.
+      Please visit the settings for this graph at https://engine.apollographql.com/ to obtain an API key for a service.
+      `);
```

Comment:
Still generally a concern about the first line of the error message being blank for the reasons mentioned in a comment just above.